cmds/free/free.go

```diff
@@ -0,0 +1,343 @@
+		value>>shift,
+		decimal,
+		units[shift/10],
+	)
+}
+
+// formatValueByConfig formats a size in bytes in the appropriate unit,
+// depending on whether FreeConfig specifies a human-readable format or a
+// specific unit
+func formatValueByConfig(value uint64, config *FreeConfig) string {
+	if config.HumanOutput {
+		return humanReadableValue(value)
+	}
+	// units and decimal part are not printed when a unit is explicitly specified
+	return fmt.Sprintf("%v", value>>config.Unit)
+}
+
+// getMainMemInfo prints the physical memory information in the specified units. Only
+// the relevant fields will be used from the input map.
+func getMainMemInfo(m meminfomap, config *FreeConfig) (*mainMemInfo, error) {
+	fields := []string{
+		"MemTotal",
+		"MemFree",
+		"Buffers",
+		"Cached",
+		"Shmem",
+		"SReclaimable",
+		"MemAvailable",
+	}
+	if missingRequiredFields(m, fields) {
+		return nil, fmt.Errorf("Missing required fields from meminfo")
+	}
+
+	// These values are expressed in kibibytes, convert to the desired unit
+	memTotal := m["MemTotal"] << KB
+	memFree := m["MemFree"] << KB
+	memShared := m["Shmem"] << KB
+	memCached := (m["Cached"] + m["SReclaimable"]) << KB
+	memBuffers := (m["Buffers"]) << KB
+	memUsed := memTotal - memFree - memCached - memBuffers
+	if memUsed < 0 {
+		memUsed = memTotal - memFree
+	}
+	memAvailable := m["MemAvailable"] << KB
+
+	mmi := mainMemInfo{
+		Total:     memTotal,
+		Used:      memUsed,
+		Free:      memFree,
+		Shared:    memShared,
+		Cached:    memCached,
+		Buffers:   memBuffers,
+		Available: memAvailable,
+	}
+	return &mmi, nil
+}
+
+// getSwapInfo prints the swap space information in the specified units. Only the
+// relevant fields will be used from the input map.
+func getSwapInfo(m meminfomap, config *FreeConfig) (*swapInfo, error) {
+	fields := []string{
+		"SwapTotal",
+		"SwapFree",
+	}
+	if missingRequiredFields(m, fields) {
+		return nil, fmt.Errorf("Missing required fields from meminfo")
+	}
+	// These values are expressed in kibibytes, convert to the desired unit
+	swapTotal := m["SwapTotal"] << KB
+	swapUsed := (m["SwapTotal"] - m["SwapFree"]) << KB
+	swapFree := m["SwapFree"] << KB
+
+	si := swapInfo{
+		Total: swapTotal,
+		Used:  swapUsed,
+		Free:  swapFree,
+	}
+	return &si, nil
+}
+
+// Free prints physical memory and swap space information. The fields will be
+// expressed with the specified unit (e.g. KB, MB)
+func Free(config *FreeConfig) error {
+	m, err := meminfo()
+	if err != nil {
+		log.Fatal(err)
+	}
+	mmi, err := getMainMemInfo(m, config)
+	if err != nil {
+		return err
+	}
+	si, err := getSwapInfo(m, config)
+	if err != nil {
+		return err
+	}
+	mi := MemInfo{Mem: *mmi, Swap: *si}
+	if config.ToJSON {
+		jsonData, err := json.Marshal(mi)
+		if err != nil {
+			return err
+		}
+		fmt.Println(string(jsonData))
+	} else {
+		fmt.Printf("              total        used        free      shared  buff/cache   available\n")
+		fmt.Printf("%-7s %11v %11v %11v %11v %11v %11v\n",
```

Comment:
Will the formatting still look good if there are more than 11 digits (12 digits = 100,000,000,000B = 100GB)?

Reply:
it will be unaligned, but that's something accepted and documented in `free`'s man page, so I am just using the same approach here. From `free(1)`:

```
If unit is missing, and you have exbibyte of RAM or swap, the number is in tebibytes and columns might not be aligned with header.
```

Reply:
yeah!